gadget/device_linux.go

```diff
@@ -239,3 +271,31 @@ func ParentDiskFromPartition(partition string) (string, error) {
 	}
 	return mainDev, nil
 }
+
+func resolveParentOfDeviceMapperDevice(devname string) (string, error) {
+	// devices set up by device mapper have /dev/block/dm-*/slaves/ which
+	// lists the devices that are upper in the chain, follow that to find
+	// the first device that is non-dm one
+	dmSlavesLevel := 0
+	const maxDmSlavesLevel = 5
+	for strings.HasPrefix(devname, "dm-") {
```

Comment:
Are we iterating for the case where we have e.g. LUKS inside LVM?

Reply:
Yes, this should be fine for more complex setups, like LVM in LUKS (checked the code for resolving the device in such setup locally) or other configurations.

Reply:
If you have a similar setup locally, you can run this to check that the device is identified correctly:
```
package main

import (
        "fmt"
        "log"
        "os"

        "github.com/snapcore/snapd/gadget"
)

func main() {
        disk, err := gadget.ParentDiskFromMountSource(os.Args[1])
        if err != nil {
                log.Fatalf("failed: %v", err)
        }
        fmt.Printf("device: %v\n", disk)
}
```